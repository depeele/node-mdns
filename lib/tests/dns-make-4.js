var Util    = require('util'),
    Utils   = require('../utils.js'),
    Message = require('../message.js'),

    expect  = [ // Header
                0x12, 0x34, 0x80, 0x80, 0x00, 0x01, 0x00, 0x03,
                0x00, 0x04, 0x00, 0x05,

                // Message
                0x04, 0x74, 0x65, 0x73, 0x74, 0x05, 0x6c, 0x6f,
                0x63, 0x61, 0x6c, 0x00, 0x01, 0x00, 0x01, 0xc0,
                0x0c, 0x00, 0x06, 0x00, 0x01, 0x00, 0x00, 0x00,
                0x01, 0x00, 0x23, 0x0a, 0x68, 0x6f, 0x73, 0x74,
                0x6d, 0x61, 0x73, 0x74, 0x65, 0x72, 0xc0, 0x0c,
                0xc0, 0x27, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00,
                0x00, 0x02, 0x00, 0x00, 0x00, 0x03, 0x00, 0x00,
                0x00, 0x04, 0x00, 0x00, 0x00, 0x05, 0xc0, 0x0c,
                0x00, 0x10, 0x00, 0x01, 0x00, 0x00, 0x00, 0x02,
                0x00, 0x0c, 0x0b, 0x48, 0x65, 0x6c, 0x6c, 0x6f,
                0x20, 0x57, 0x6f, 0x72, 0x6c, 0x64, 0xc0, 0x0c,
                0x00, 0x0f, 0x00, 0x01, 0x00, 0x00, 0x00, 0x03,
                0x00, 0x09, 0x00, 0x0a, 0x04, 0x6d, 0x61, 0x69,
                0x6c, 0xc0, 0x0c, 0xc0, 0x0c, 0x00, 0x02, 0x00,
                0x01, 0x00, 0x00, 0x00, 0x04, 0x00, 0x06, 0x03,
                0x6e, 0x73, 0x31, 0xc0, 0x0c, 0xc0, 0x0c, 0x00,
                0x02, 0x00, 0x01, 0x00, 0x00, 0x00, 0x05, 0x00,
                0x06, 0x03, 0x6e, 0x73, 0x32, 0xc0, 0x0c, 0xc0,
                0x0c, 0x00, 0x02, 0x00, 0x01, 0x00, 0x00, 0x00,
                0x06, 0x00, 0x06, 0x03, 0x6e, 0x73, 0x33, 0xc0,
                0x0c, 0xc0, 0x0c, 0x00, 0x02, 0x00, 0x01, 0x00,
                0x00, 0x00, 0x07, 0x00, 0x06, 0x03, 0x6e, 0x73,
                0x34, 0xc0, 0x0c, 0xc0, 0x70, 0x00, 0x01, 0x00,
                0x01, 0x00, 0x00, 0x00, 0x08, 0x00, 0x04, 0x7f,
                0x00, 0x00, 0x01, 0xc0, 0x83, 0x00, 0x01, 0x00,
                0x01, 0x00, 0x00, 0x00, 0x09, 0x00, 0x04, 0x7f,
                0x00, 0x00, 0x01, 0xc0, 0x95, 0x00, 0x01, 0x00,
                0x01, 0x00, 0x00, 0x00, 0x0a, 0x00, 0x04, 0x7f,
                0x00, 0x00, 0x02, 0xc0, 0xa7, 0x00, 0x01, 0x00,
                0x01, 0x00, 0x00, 0x00, 0x0b, 0x00, 0x04, 0x7f,
                0x00, 0x00, 0x03, 0xc0, 0xb9, 0x00, 0x01, 0x00,
                0x01, 0x00, 0x00, 0x00, 0x0c, 0x00, 0x04, 0x7f,
                0x00, 0x00, 0x04
    ];

/*****************************************************************************
 * Creation test/pack
 *
 */
console.log("--------------------------------------------------------------");
var message = new Message( ),
    name    = 'test.local';

message.header.id      = 0x1234;
message.header.qr      = 1;
message.header.ra      = 1;         // authoritative
message.header.rd      = 0;

message.addQuestion(name, 1, 1);

message.header.anCount = 3;
message.header.nsCount = 4;
message.header.arCount = 5;

message.addRR('answer', name, 'SOA', 'IN', 1, {
    mname:      'hostmaster.'+ name,
    rname:      'hostmaster.'+ name,
    serial:     1,
    refresh:    2,
    retry:      3,
    expire:     4,
    minimum:    5
});
message.addRR('answer', name, 'TXT', 'IN', 2, {txt: ['Hello World']} );
message.addRR('answer', name, 'MX',  'IN', 3, {preference: 10, exchange: 'mail.'+ name});

message.addRR('authority', name, 'NS',  'IN', 4, {ns: 'ns1.'+ name});
message.addRR('authority', name, 'NS',  'IN', 5, {ns: 'ns2.'+ name});
message.addRR('authority', name, 'NS',  'IN', 6, {ns: 'ns3.'+ name});
message.addRR('authority', name, 'NS',  'IN', 7, {ns: 'ns4.'+ name});

message.addRR('additional', 'mail.'+ name, 'A', 'IN',  8, {a: '127.0.0.1'});
message.addRR('additional', 'ns1.'+ name,  'A', 'IN',  9, {a: '127.0.0.1'});
message.addRR('additional', 'ns2.'+ name,  'A', 'IN', 10, {a: '127.0.0.2'});
message.addRR('additional', 'ns3.'+ name,  'A', 'IN', 11, {a: '127.0.0.3'});
message.addRR('additional', 'ns4.'+ name,  'A', 'IN', 12, {a: '127.0.0.4'});


console.log("Generated message:\n%s", message);

//console.log("%s", Util.inspect(message, false, 20));

console.log("--------------------------------------------------------------");
var wBuf    = new Buffer( 4096 );

message.pack(wBuf);

console.log("Packed message (%d bytes):\n%s",
            message.produced, Utils.buf2hex(wBuf, {
                                            length:     message.produced,
                                            octetsPer:  16
                             }));
console.log(">>> Expect (%d bytes):\n%s",
            expect.length, Utils.buf2hex(expect, {octetsPer:16 }));

if (message.produced !== expect.length)
{
    console.log("*** Pack error: length %d != expected %d",
                message.produced, expect.length);
    process.exit(-1);
}

var diff    = 0;
for (var idex = 0, len = expect.length; idex < len; idex++)
{
    if (wBuf[idex] !== expect[idex])
    {
        diff++;
        console.log("*** Pack error: @ %s: %s != expected %s",
                    Utils.int2hex(idex, 2),
                    Utils.int2hex(wBuf[idex]),
                    Utils.int2hex(expect[idex]));

        if (diff > 8)   { break; }
    }
}

if (diff === 0)
{
    console.log(">>> packed buffer matches original source");
}
