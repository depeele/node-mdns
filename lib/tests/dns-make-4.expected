--------------------------------------------------------------
Generated message:
Header: id: 4660, opcode: QUERY, status: NOERROR
      : flags: qr ra
      : Question: 1, Answer: 3, Authority: 4, Additional: 5

1 question
   0: test.local: qtype:1 (A), qclass:1 (IN)

3 answers
   0: test.local: type:6 (SOA), class:1 (IN), ttl:1
        rdata: mname:hostmaster.test.local, rname:hostmaster.test.local, 
             : serial:1, refresh:2, retry:3, expire:4, minimum:5

   1: test.local: type:16 (TXT), class:1 (IN), ttl:2
        rdata: txt:Hello World

   2: test.local: type:15 (MX), class:1 (IN), ttl:3
        rdata: preference:10, exchange:mail.test.local


4 authorities
   0: test.local: type:2 (NS), class:1 (IN), ttl:4
        rdata: ns:ns1.test.local

   1: test.local: type:2 (NS), class:1 (IN), ttl:5
        rdata: ns:ns2.test.local

   2: test.local: type:2 (NS), class:1 (IN), ttl:6
        rdata: ns:ns3.test.local

   3: test.local: type:2 (NS), class:1 (IN), ttl:7
        rdata: ns:ns4.test.local


5 additionals
   0: mail.test.local: type:1 (A), class:1 (IN), ttl:8
        rdata: a:127.0.0.1

   1: ns1.test.local: type:1 (A), class:1 (IN), ttl:9
        rdata: a:127.0.0.1

   2: ns2.test.local: type:1 (A), class:1 (IN), ttl:10
        rdata: a:127.0.0.2

   3: ns3.test.local: type:1 (A), class:1 (IN), ttl:11
        rdata: a:127.0.0.3

   4: ns4.test.local: type:1 (A), class:1 (IN), ttl:12
        rdata: a:127.0.0.4


--------------------------------------------------------------
Packed message (271 bytes):
 0x0000: 12 34 80 80 00 01 00 03 00 04 00 05 04 74 65 73 
 0x0010: 74 05 6c 6f 63 61 6c 00 01 00 01 c0 0c 00 06 00 
 0x0020: 01 00 00 00 01 00 23 0a 68 6f 73 74 6d 61 73 74 
 0x0030: 65 72 c0 0c c0 27 00 00 00 01 00 00 00 02 00 00 
 0x0040: 00 03 00 00 00 04 00 00 00 05 c0 0c 00 10 00 01 
 0x0050: 00 00 00 02 00 0c 0b 48 65 6c 6c 6f 20 57 6f 72 
 0x0060: 6c 64 c0 0c 00 0f 00 01 00 00 00 03 00 09 00 0a 
 0x0070: 04 6d 61 69 6c c0 0c c0 0c 00 02 00 01 00 00 00 
 0x0080: 04 00 06 03 6e 73 31 c0 0c c0 0c 00 02 00 01 00 
 0x0090: 00 00 05 00 06 03 6e 73 32 c0 0c c0 0c 00 02 00 
 0x00a0: 01 00 00 00 06 00 06 03 6e 73 33 c0 0c c0 0c 00 
 0x00b0: 02 00 01 00 00 00 07 00 06 03 6e 73 34 c0 0c c0 
 0x00c0: 70 00 01 00 01 00 00 00 08 00 04 7f 00 00 01 c0 
 0x00d0: 83 00 01 00 01 00 00 00 09 00 04 7f 00 00 01 c0 
 0x00e0: 95 00 01 00 01 00 00 00 0a 00 04 7f 00 00 02 c0 
 0x00f0: a7 00 01 00 01 00 00 00 0b 00 04 7f 00 00 03 c0 
 0x0100: b9 00 01 00 01 00 00 00 0c 00 04 7f 00 00 04
>>> Expect (271 bytes):
 0x0000: 12 34 80 80 00 01 00 03 00 04 00 05 04 74 65 73 
 0x0010: 74 05 6c 6f 63 61 6c 00 01 00 01 c0 0c 00 06 00 
 0x0020: 01 00 00 00 01 00 23 0a 68 6f 73 74 6d 61 73 74 
 0x0030: 65 72 c0 0c c0 27 00 00 00 01 00 00 00 02 00 00 
 0x0040: 00 03 00 00 00 04 00 00 00 05 c0 0c 00 10 00 01 
 0x0050: 00 00 00 02 00 0c 0b 48 65 6c 6c 6f 20 57 6f 72 
 0x0060: 6c 64 c0 0c 00 0f 00 01 00 00 00 03 00 09 00 0a 
 0x0070: 04 6d 61 69 6c c0 0c c0 0c 00 02 00 01 00 00 00 
 0x0080: 04 00 06 03 6e 73 31 c0 0c c0 0c 00 02 00 01 00 
 0x0090: 00 00 05 00 06 03 6e 73 32 c0 0c c0 0c 00 02 00 
 0x00a0: 01 00 00 00 06 00 06 03 6e 73 33 c0 0c c0 0c 00 
 0x00b0: 02 00 01 00 00 00 07 00 06 03 6e 73 34 c0 0c c0 
 0x00c0: 70 00 01 00 01 00 00 00 08 00 04 7f 00 00 01 c0 
 0x00d0: 83 00 01 00 01 00 00 00 09 00 04 7f 00 00 01 c0 
 0x00e0: 95 00 01 00 01 00 00 00 0a 00 04 7f 00 00 02 c0 
 0x00f0: a7 00 01 00 01 00 00 00 0b 00 04 7f 00 00 03 c0 
 0x0100: b9 00 01 00 01 00 00 00 0c 00 04 7f 00 00 04
>>> packed buffer matches original source
